File: Q2/Q2_Matrix_Multiplication.cpp
#include <iostream>
#include <time.h>

using namespace std;

#define N 512

long current_time()
{
    struct timespec time;
    timespec_get(&time, CLOCK_MONOTONIC);
    return time.tv_sec * 1000000000 + time.tv_nsec;
}

void multiply(double matrix1[N][N],
              double matrix2[N][N],
              double result[N][N])
{
    int i, j, k;
    for (i = 0; i < N; i++)
    {
        for (j = 0; j < N; j++)
        {
            result[i][j] = 0;
            for (k = 0; k < N; k++)
                result[i][j] += matrix1[i][k] * matrix2[k][j];
        }
    }
}

void random_matrix(double matrix[N][N])
{
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < N; j++)
        {
            matrix[i][j] = (rand() % 100) * 0.21;
        }
    }
}

int main()
{
    int i, j;
    double result[N][N];
    double matrix1[N][N];
    double matrix2[N][N];
    random_matrix(matrix1);
    random_matrix(matrix2);
    srand(time(0));

    long start_time = current_time();

    multiply(matrix1, matrix2, result);

    long end_time = current_time();

    cout << "Multiplied matrix is \n";
    for (i = 0; i < N; i++)
    {
        cout << "[";
        for (j = 0; j < N; j++)
        {
            cout << result[i][j] << " ";
        }
        cout << "]"
             << "\n";
    }

    cout << endl
         << "Total time taken in ns = " << end_time - start_time << endl;

    return 0;
}

File: Q2/Q2_Matrix_Multiplication.py
import random
import timeit

global n
n = 512

def print_random_matrix(n, matrix):
    for i in range(n):
        for j in range(n):     
            print (matrix[i][j], end = " ")       

        print()


def multiply(n, matrix1, matrix2):
 
    result = [[0 for x in range(n)]
              for y in range (n)]
     
    for i in range(n):
        for j in range(n):
            result[i][j] = 0
            for x in range(n):           
                result[i][j] += (matrix1[i][x] * matrix2[x][j])
            
    for i in range(n):
        for j in range(n):     
            print (result[i][j], end = " ")       
        print()


if __name__ == "__main__":

    # there is no double in python, hence using float 
    matrix1 = [[float(random.randint(0,15) + random.random()) for x in range(n)]
              for y in range (n)]

    matrix2 = [[float(random.randint(0,15) + random.random()) for x in range(n)]
              for y in range (n)]

    print_random_matrix(n, matrix1)
    print()
    print_random_matrix(n, matrix2)
    print()
    t = timeit.timeit(lambda: multiply(n, matrix1, matrix2), number = 1)

    print("Time in sec = ",t)
